src/Hamiltonians/abstract.jl

```diff
@@ -144,47 +142,43 @@ julia> rayleigh_quotient(mom, v) # momentum expectation value for state vector `
 momentum
 
 """
-    sm, basis = build_sparse_matrix_from_LO(
-        ham, address=starting_address(ham); cutoff, filter=nothing, nnzs, col_hint, sort=false, kwargs...
-    )
+    build_sparse_matrix_from_LO(
+        ham, address=starting_address(ham);
+        cutoff, filter=nothing, nnzs, col_hint, sort=false, kwargs...
+    ) -> sm, basis
+    build_sparse_matrix_from_LO(ham, addresses::AbstractVector; kwargs...)
 
 Create a sparse matrix `sm` of all reachable matrix elements of a linear operator `ham`
-starting from `address`. The vector `basis` contains the addresses of basis
-configurations.
-
-Providing the number `nnzs` of expected calculated matrix elements may improve performance.
-The default estimates for `nnzs` is `dimension(ham)`.
+starting from `address`. Instead of a single address, a vector of `addresses` can be passed.
+The vector `basis` contains the addresses of basis configurations.
 
-Setting a custom value `col_hint` for the estimated number of nonzero 
-off-diagonal matrix elements in each matrix column may improve performance.
-The default value for `col_hint` is `num_offdiagonals(ham, address)`.
+Providing the number `nnzs` of expected calculated matrix elements and `col_hint` for the
+estimated number of nonzero off-diagonal matrix elements in each matrix column may improve
+performance.
 
 Providing an energy cutoff will skip the columns and rows with diagonal elements greater
 than `cutoff`. Alternatively, an arbitrary `filter` function can be used instead. These are
-not enabled by default.
+not enabled by default. To generate the matrix truncated to the subspace spanned by the
+`addresses`, use `filter = Returns(false)`.
 
-Setting `sort` to `true` will sort the matrix rows and columns. This is useful when the
-order of the columns matters, e.g. when comparing matrices. Any additional keyword arguments
-are passed on to `Base.sortperm`.
+Setting `sort` to `true` will sort the `basis` and order the matrix rows and columns
+accordingly. This is useful when the order of the columns matters, e.g. when comparing
+matrices. Any additional keyword arguments are passed on to `Base.sortperm`.
 
 See [`BasisSetRep`](@ref).
 """
 function build_sparse_matrix_from_LO(
-    ham, address=starting_address(ham);
+    ham, addr_or_vec=starting_address(ham);
     cutoff=nothing,
     filter=isnothing(cutoff) ? nothing : (a -> diagonal_element(ham, a) ≤ cutoff),
-    nnzs=dimension(ham), col_hint=num_offdiagonals(ham, address),
-    sort=false, kwargs...,
+    nnzs=0, col_hint=0, # sizehints are opt-in
+    sort=false, kwargs...
 )
-    if !isnothing(filter) && !filter(address)
-        throw(ArgumentError(string(
-            "Starting address does not pass `filter`. ",
-            "Please pick a different address or a different filter."
-        )))
-    end
```

Comment:
Why was the filter on the input address removed?

Reply:
The main reason was to enable truncating the matrix by passing a simple filter like `filter = Returns(false)`. If we keep checking the input address, this would just error and not do anything.

The change means that that the code now assumes that by passing an address the user is committed to use that address as part of the truncated subspace. This makes more sense to me than the old behaviour.